File: Convert-a-string-to-a-number/string-to-number.php
<?php
// We need a function that can transform a string into a number. What ways of achieving this do you know?
//Note: Don't worry, all inputs will be strings, and every string is a perfectly valid representation of an integral number.

// Examples
// "1234"  --> 1234
// "605"   --> 605
// "1405"  --> 1405
// "-7"    --> -7


// solution 1: use php's explicit type casting
// Note: generally faster than using built-in intval() function, but cannot be used in callback functions like array_map()

function stringToNumber($str) {
  return (int)$str;
}

// Examples
echo (int)"123";   // 123
echo (int)"-456";  // -456
echo (int)"0";     // 0


//solution 2: build the number digit by digit (Brute Force)
function stringToNumbers($str){
  $result = 0;
  $isNegative = False;
  $startIndex = 1;

  //check for negative sign
  if ($str[0] === '-') {
    $isNegative = True;
    $startIndex = 1;
  }

  //process each digit
  for ($i = $startIndex; $i < strlen($str); $i++) {
    $digit = ord($str[$i]) - ord('0');
    $result = $result * 10 + $digit;
  }
  return $isNegative ? -$result : $result;
} 
?>

File: Convert-a-string-to-a-number/string_to_number.py
# We need a function that can transform a string into a number. What ways of achieving this do you know?
# Note: Don't worry, all inputs will be strings, and every string is a perfectly valid representation of an integral number.

# Examples
# "1234"  --> 1234
# "605"   --> 605
# "1405"  --> 1405
# "-7"    --> -7


# solution 1: use pythons's built-in parse function
def string_to_number(s) :
  return int(s)

#Examples
print(int("1234"))  # 1234
print(int("605"))   # 605
print(int("1405"))  # 1405
print(int("-7"))    # -7

# Solution 2: Brute force
def string_to_numbers(s) :
  result = 0
  is_negative = False
  start_index = 0

  # check for negative sign
  if s[0] == '-':
    is_negative = True
    start_index = 1
  
  # Process each digit
  for i in range(start_index, len(s)):
    digit = ord(s[i]) - ord('0') # convert char to digit
    result = result * 10 + digit

  return -result if is_negative else result